src/gt4py/next/ffront/ast_passes/single_static_assign.py

```diff
@@ -11,12 +11,77 @@
 # distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
 #
 # SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
 
 import ast
+import dataclasses
+import typing
 
 from gt4py.next.ffront.fbuiltins import TYPE_BUILTIN_NAMES
 
 
+ASTNodeT = typing.TypeVar("ASTNodeT", bound=ast.AST)
+
+_UNIQUE_NAME_SEPERATOR = "ᐞ"
+
+
+def unique_name(name: str, num_assignments: int):
+    """Given the name of a variable return a unique name after the given amount of assignments."""
+    if num_assignments >= 0:
+        return f"{name}{_UNIQUE_NAME_SEPERATOR}{num_assignments}"
+    assert num_assignments == -1
+    return name
+
+
+def _make_assign(target: str, source: str, location_node: ast.AST) -> ast.Assign:
+    result = ast.Assign(
+        targets=[ast.Name(ctx=ast.Store(), id=target)], value=ast.Name(ctx=ast.Load(), id=source)
+    )
+    for node in ast.walk(result):
+        ast.copy_location(node, location_node)
+    return result
+
+
+def _is_guaranteed_to_return(node: ast.stmt | list[ast.stmt]) -> bool:
+    if isinstance(node, list):
+        return any(_is_guaranteed_to_return(child) for child in node)
+    if isinstance(node, ast.Return):
+        return True
+    if isinstance(node, ast.If):
+        return _is_guaranteed_to_return(node.body) and _is_guaranteed_to_return(node.orelse)
+    return False
+
+
+@dataclasses.dataclass
+class _AssignmentTracker:
+    """Helper class to keep track of the number of assignments to a variable."""
+
+    _counts: dict[str, int] = dataclasses.field(default_factory=dict)
+
+    def define(self, name: str) -> None:
+        if name in self.names():
+            raise ValueError(f"Variable {name} is already defined.")
+        self._counts[name] = -1
+
+    def assign(self, name: str) -> None:
+        self._counts[name] = self.count(name) + 1
+
+    def count(self, name: str):
+        return self._counts.get(name, -1)
+
+    def names(self):
+        return self._counts.keys()
+
+    def copy(self):
+        return _AssignmentTracker({**self._counts})
+
+
+def _merge_assignment_tracker(a: _AssignmentTracker, b: _AssignmentTracker) -> _AssignmentTracker:
+    common_names = set(a.names()) & set(b.names())
```

Comment:
The conversion to `set` is not needed since `dict_keys` is already a `collection.abc.Set`:
```suggestion
    common_names = a.names() & b.names()
```